rosdoc2/verbs/build/builders/sphinx_builder.py

```diff
@@ -366,10 +409,55 @@ def build(self, *, doc_build_folder, output_staging_directory):
             if package_name != self.build_context.package.name
         ]
 
+        # If the package has build type `ament_python`, or if the user configured
+        # to run `sphinx-apidoc`, then invoke `sphinx-apidoc` before building
+        if (
+            self.build_context.build_type == 'ament_python' or
+            self.build_context.always_run_sphinx_apidoc
+        ):
+            package_xml_directory = os.path.dirname(self.build_context.package.filename)
+            # If 'python_source' is specified, construct 'package_src_directory' from it
+            if self.build_context.python_source is not None:
+                package_src_directory = \
+                    os.path.join(
+                        package_xml_directory,
+                        self.build_context.python_source)
+            # If not provided, try to find the package source direcotry
+            else:
+                package_list = setuptools.find_packages(where=package_xml_directory)
+                if self.build_context.package.name in package_list:
+                    package_src_directory = os.path.join(
+                        package_xml_directory,
+                        self.build_context.package.name)
+                else:
+                    package_src_directory = None
+
+            if not package_src_directory or not os.path.isdir(package_src_directory):
+                raise RuntimeError(
+                    "Could not locate source directory to invoke sphinx-apidoc in. "
+                    "If this is package does not have a standard Python package layout, "
+                    "please specify the Python source in 'rosdoc2.yaml'.")
+            cmd = [
+                'sphinx-apidoc',
+                '-o', os.path.relpath(sourcedir, start=doc_build_folder),
+                '-e',  # Document each module in its own page.
+                os.path.abspath(package_src_directory),
+            ]
+            logger.info(
+                f"Running sphinx-apidoc: '{' '.join(cmd)}' in '{doc_build_folder}'"
+            )
+            completed_process = subprocess.run(cmd, cwd=doc_build_folder)
+            msg = f"sphinx-apidoc exited with return code '{completed_process.returncode}'"
+            if completed_process.returncode == 0:
+                logger.info(msg)
+            else:
+                raise RuntimeError(msg)
+
         # Setup rosdoc2 Sphinx file which will include and extend the one in `sourcedir`.
         self.generate_wrapping_rosdoc2_sphinx_project_into_directory(
             doc_build_folder,
             sourcedir,
+            package_src_directory,
```

Comment:
Here is where package_src_directory is used, outside of the if statement. This is also where my demo package fails "local variable 'package_src_directory' referenced before assignment"

Reply:
Sorry to trickle in these comments like this. As written this did not work fort me (after I fixed other issues reported earlier.) Here is what works:

        # Setup rosdoc2 Sphinx file which will include and extend the one in `sourcedir`.
        self.generate_wrapping_rosdoc2_sphinx_project_into_directory(
            doc_build_folder,
            sourcedir,
            #package_src_directory,
            os.path.abspath(package_src_directory),
            intersphinx_mapping_extensions)

Reply:
Thank you for the fix! Addressed in https://github.com/ros-infrastructure/rosdoc2/pull/28/commits/45b8e08a534df0c8bc9003acc5db32e9da92f374.